Migrate cache api intro doc (#2643)
GoogleChrome/web.dev: src/site/content/en/blog/cache-api-quick-guide/index.md

## What can be stored

The caches only store pairs of [`Request`][mdn-request] and
[`Response`][mdn-response] objects, representing HTTP requests and responses,
respectively. However, the requests and responses can contain any kind of data
that can be transferred over HTTP.

### How much can be stored?

In short, **a lot**, at least a couple of hundred megabytes, and potentially
hundreds of gigabytes or more. Browser implementations vary, but the amount
of storage available is usually based on the amount of storage available on
the device.

## Creating and opening a cache

To open a cache, use the `caches.open(name)` method, passing the name of the
cache as the single parameter. If the named cache does not exist, it is
created. This method returns a `Promise` that resolves with the `Cache` object.

```js
const cache = await caches.open('my-cache');
// do something with cache...
```

## Adding to a cache

There are three ways to add an item to a cache - `add`, `addAll`, and `put`.
All three methods return a `Promise`.

### `cache.add`

First, there is `cache.add()`. It takes one parameter, either a `Request`

> or a URL (`string`). It makes a request to the network and stores to response

Reviewer: ```suggestion
or a URL (`string`). It makes a request to the network and stores the response
```

in the cache. If the
fetch fails, or if the status code of the response is not in the 200 range,
then nothing is stored and the `Promise` rejects. Note that cross-origin
requests not in CORS mode cannot be stored because they return a `status` of

> `0`, such requests can only be stored with `put`.

Reviewer: ```suggestion
`0`. Such requests can only be stored with `put`.
```

```js
// Retreive data.json from the server and store the response.
cache.add(new Request('/data.json'));

// Retreive data.json from the server and store the response.
cache.add('/data.json');
```

### `cache.addAll`

Next, there is `cache.addAll()`. It works similarly to `add()`, but takes an

> array of `Request` objects or URLs (`strings`). This works similarly to

Reviewer: ```suggestion
array of `Request` objects or URLs (`string`s). This works similarly to
```

calling `cache.add` for each individual request, except that the `Promise`
rejects if any single request is not cached.

```js
const urls = ['/weather/today.json', '/weather/tomorrow.json'];
cache.addAll(urls);
```

In each of these cases, a new entry overwrites any matching existing entry.
This uses the same matching rules described in the section on

> [retrieving](#retreiving).

Reviewer: ```suggestion
[retrieving](#retrieving).
```
Typo fix

    </tr>
    <tr>
      <td><code>json</code></td>
      <td>
        Interprets the bytes of the body as a UTF-8 encoded string, then tries
        to parse it as JSON. Returns the resulting object, or throws a
        <code>TypeError</code> if the string cannot be parsed as JSON.
      </td>
    </tr>
    <tr>
      <td><code>formData</code></td>
      <td>
        Interprets the bytes of the body as an HTML form, encoded either as
        <code>multipart/form-data</code> or
        <code>application/x-www-form-urlencoded</code>. Returns a
        <a href="https://developer.mozilla.org/en-US/docs/Web/API/FormData">FormData</a>
        object, or throws a <code>TypeError</code> if the data cannot be parsed.
      </td>
    </tr>
    <tr>
      <td><code>body</code></td>
      <td>
        Returns a <a href="https://developer.mozilla.org/en-US/docs/Web/API/ReadableStream">ReadableStream</a>
        for the body data.
      </td>
    </tr>
  </tbody>
</table>

For example

```js
const response = new Response('Hello world');
const buffer = await response.arrayBuffer();
console.log(new Uint8Array(buffer));
// Uint8Array(11) [72, 101, 108, 108, 111, 32, 119, 111, 114, 108, 100]
```

> ## Retrieving from a cache {: #retreiving }

Reviewer: ```suggestion
## Retrieving from a cache {: #retrieving }
```
Typo fix

To find an item in a cache, you can use the `match` method.

```js
const response = await cache.match(request);
console.log(request, response);
```

> If `request` is a string it is first be converted to a `Request` by calling

Reviewer: ```suggestion
If `request` is a string it first needs to be converted to a `Request` by calling
```
@petele I'm unsure if you meant that the browser converts this or if it's something the developer needs to do

Author: Good catch - the browser does it. Will fix.

`new Request(request)`. The function returns a `Promise` that resolves to
a `Response` if a matching entry is found, or `undefined` otherwise.

> To determine if two `Requests` match, more than just the URL is used. Two

Reviewer: ```suggestion
To determine if two `Requests` match, the browser uses more than just the URL. Two
```
I think we're talking about the browser's logic but it's not clear to me because we don't specifically mention the subject of the sentence

requests are considered different if they have different query strings,

> `Vary` headers and/or methods (`GET`, `POST`, `PUT`, etc.).

Reviewer: ```suggestion
`Vary` headers, or HTTP methods (`GET`, `POST`, `PUT`, etc.).
```
I think just using `or` and removing `and` is more appropriate here because we're talking about boolean "any" logic (if any of the conditions are true, then the 2 requests are considered different)

You can ignore some or all of these things by passing an options object as a
second parameter.

```js
const options = {
  ignoreSearch: true,
  ignoreMethod: true,
  ignoreVary: true
};

const response = await cache.match(request, options);
// do something with the response
```

If more than one cached request matches then the one that was created first is
returned. If you want to retrieve *all* matching responses, you can use
`cache.matchAll()`.

```js
const options = {
  ignoreSearch: true,
  ignoreMethod: true,
  ignoreVary: true
};

const responses = await cache.matchAll(request, options);
console.log(`There are ${responses.length} matching responses.`);
```

As a shortcut you can search over all caches at once by using `caches.match()`
instead of calling `cache.match()` for each cache.

## Searching

The Cache API does not provide a way to search for requests or responses
except for matching entries against a `Response` object. However, you can
implement your own search using filtering or by creating an index.

### Filtering

One way to implement your own search is to iterate over all entries and
filter down to the ones that you want. Let's say that you want to find all

> items that have URLs ending with '.png'.

Reviewer: ```suggestion
items that have URLs ending with `.png`.
```

```js
async function findImages() {
  // Get a list of all of the caches for this origin
  const cacheNames = await caches.keys();
  const result = [];

  for (const name of cacheNames) {
    // Open the cache
    const cache = await caches.open(name);

    // Get a list of entries. Each item is a Request object
    for (const request of await cache.keys()) {
      // If the request URL matches, add the response to the result
      if (request.url.endsWith('.png')) {
        result.push(await cache.match(request));
      }
    }
  }

  return result;
}
```

This way you can use any property of the `Request` and `Response` objects to
filter the entries. Note that this is slow if you search over large sets of
data.

### Creating an index

The other way to implement your own search is to maintain a separate index of

> entries that can be searched, stored in IndexedDB. Since this is the kind of

Reviewer: ```suggestion
entries that can be searched and store the index in IndexedDB. Since this is the kind of
```
Suggested rephrasing because I stumbled over this sentence and it took me a couple reads to figure out what it was saying
